File: Code/1456/1456.cpp
/******************************************************************************
 * 文件名: 1456
 * 作者: BOLUN XU
 * 创建日期: 2025
 * 版本: 1.0
 * 描述: 定长子串中元音的最大数目。
 * 时间复杂度： O(N)
 * 空间复杂度： O(1)
 * 执行用时： 3 ms
 * 消耗内存： 13.00 mb
 ******************************************************************************/
#include <iostream>

using namespace std;

class Solution {
public:
  int maxVowels(string s, int k) {
    int ans = 0;
    int vow = 0;
    for (int i = 0; i < s.length(); i++) {
      // 进入窗口
      // cout << "DEBUG: ----------------------------------------" << endl;
      // cout << "DEBUG: s[" << i << "]= " << s[i] << " 进入窗口" << endl;
      if (s[i] == 'a' || s[i] == 'e' || s[i] == 'i' || s[i] == 'o' || s[i] == 'u') {
        // cout << "DEBUG: s[" << i << "]= " << s[i] << " 是元音" << endl;
        vow++;
      }
      if (i < k - 1) {
        // cout << "DEBUG: 滑窗长度不满足" << endl;
        continue;
      }
      // 更新答案
      // cout << "DEBUG: 当前滑窗内元音数为 " << vow << endl;
      ans = max(ans, vow);
      // 离开窗口
      // cout << "DEBUG: s[" << i - k + 1 << "]= " << s[i - k + 1] << "离开窗口" << endl;
      if (s[i - k + 1] == 'a' || s[i - k + 1] == 'e' || s[i - k + 1] == 'i' || s[i - k + 1] == 'o' || s[i - k + 1] == 'u') {
        // cout << "DEBUG: s[" << i - k + 1 << "]= " << s[i - k + 1] << " 是元音" << endl;
        vow--;
      }
    }
    return ans;
  }
};

int main() {
  Solution s;
  cout << s.maxVowels("abciiidef", 3) << endl;
}

File: Code/1456/1456.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
文件名: PersonalLeetCode
作者: Bolun Xu
创建日期: 2025
版本: 1.0
描述: 定长子串中元音的最大数目。
时间复杂度： O(N)
空间复杂度： O(1)
执行用时： 99 ms
消耗内存： 17.94 mb
"""


class Solution:
    def maxVowels(self, s: str, k: int) -> int:
        ans = 0  # 答案
        vow = 0  # 当前滑窗内的元音数
        for i, c in enumerate(s):
            # 进入窗口
            # print(f"DEBUG: ----------------------------------------")
            # print(f"DEBUG: s[{i}]= {c} 进入窗口")
            if c in "aeiou":
                # print(f"DEBUG: s[{i}]= {c} 是元音")
                vow += 1
            if i < k - 1:
                # print(f"DEBUG: 滑窗长度不满足")
                continue
            # 更新答案
            # print(f"DEBUG: 当前滑窗内元音数为 {vow}")
            ans = max(ans, vow)
            # 离开窗口
            # print(f"DEBUG: s[{i - k + 1}]= {s[i - k + 1]} 离开窗口")
            if s[i - k + 1] in "aeiou":
                # print(f"DEBUG: s[{i - k + 1}]= {s[i - k + 1]} 是元音")
                vow -= 1
        return ans


if __name__ == '__main__':
    sol = Solution()
    print(sol.maxVowels(s="abciiidef", k=3))
